Add an option 'no_nullify' to windowFunnel()

achimbab opened the pull request:
I hereby agree to the terms of the CLA available at: https://yandex.ru/legal/cla/?lang=en

Changelog category (leave one):
- New Feature

Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):

When `windowFunnel()` aggregates null events, it returns null even though there is at least one matched event.

It is the default behavior of almost aggregating functions but I want to get an exact event-level instead of null.

Detailed description / Documentation draft:

There are some data like the below table.

| dt(DateTime) | u(User) | a(EventA) | b(EventB) |
| --------- | ----- | ------- | ------- |
| 1970-01-01 09:00:01 | 1 | a1 | b1 |
| 1970-01-01 09:00:02 | 1 | a2 | b2 |
| 1970-01-01 09:00:01 | 2 | a1 | NULL |
| 1970-01-01 09:00:02 | 2 | a2 | NULL |

```SQL
CREATE TABLE funnel_test_non_null              
(                                 
    `dt` DateTime,               
    `u` int,                                                                                                                                                                      
    `a` Nullable(String),
    `b` Nullable(String)
)
ENGINE = MergeTree()
PARTITION BY dt
ORDER BY u;

insert into funnel_test_non_null values (1, 1, 'a1', 'b1') (2, 1, 'a2', 'b2');
insert into funnel_test_non_null values (1, 2, 'a1', null) (2, 2, 'a2', null);
```

```SQL
SELECT 
    u, 
    windowFunnel(86400)(dt, a = 'a1', b = 'b2') AS level
FROM funnel_test_non_null
GROUP BY u
ORDER BY u ASC
```

The result of the `user 2` is NULL even though `user 2` has the first `a = 'a1'` event.
```
u       level
1       2
2       \N
```

So I added an option `no_nullify` to get an exact result.

```SQL
SELECT 
    u, 
    windowFunnel(86400, 'no_nullify')(dt, a = 'a1', b = 'b2') AS level
FROM funnel_test_non_null
GROUP BY u
ORDER BY u ASC
FORMAT TSVWithNames
```

With `no_nullify` option, it is able to get the exact level of `user 2`.

```
u       level
1       2
2       1
```

Thank you.

Commit:
Add a parameter called 'no_nullify' to windowFunnel

alexey-milovidov commented:
```
/AggregateFunctions/CMakeFiles/clickhouse_aggregate_functions.dir/AggregateFunctionNull.cpp.o -c ../dbms/src/AggregateFunctions/AggregateFunctionNull.cpp
2020-03-24 10:08:52 /build/obj-x86_64-linux-gnu/../dbms/src/AggregateFunctions/AggregateFunctionNull.cpp:30:17: error: do not use 'compare' to test equality of strings; use the string equality operator instead [readability-string-compare,-warnings-as-errors]
2020-03-24 10:08:52             if (option.compare(expected) == 0)
2020-03-24 10:08:52                 ^~~~~~~~~~~~~~~~~~~~~~~~    ~
2020-03-24 10:08:52                 option                      expected
```

https://clickhouse-builds.s3.yandex.net/9841/5bd3fec667a3c3751360c6a061be068bffeb585c/build_log_637340941_1585045138.txt

alexey-milovidov commented:
Is there a reson to not make it by default?

alexey-milovidov commented:
Maybe if we will make it by default, simpler solution exists.

Commit:
Use the string equality operator and add ErrorCodes

Commit:
Fix a corner case of windowFunnel with 'strict_order'

achimbab commented:
@alexey-milovidov 
I modified the code to use the string equality operator instead of the comparator and fixed the corner case of `'strict_order'`.
It is good to make it by default, but I worry about something... If someone is already using `windowFunnel`, changing default behavior affects their software. So I added `'no_nullify'` as an optional parameter.

hczhcz commented:
> Returned value
>
> The maximum number of consecutive triggered conditions from the chain within the sliding time window.
> All the chains in the selection are analyzed.

A null result is more like a bug than a feature. It looks good to make it default.

Commit:
Update doc for parameters of windowFunnel

achimbab commented:
@qoega I wrote a document at https://github.com/ClickHouse/ClickHouse/pull/9841/commits/140c7f39332f88671aa3a9ba431c723fa2dffa8c

Commit:
Make 'no_nullify' options of windowFunnel as default.

achimbab commented:
@alexey-milovidov @hczhcz 
I made it as default.
If there's any other problems please let me know.
Thank you.

Commit:
Simplify if-statement in AggregateFunctionNull.cpp

alexey-milovidov commented on dbms/src/AggregateFunctions/AggregateFunctionWindowFunnelNoNullifyVariadic.h:
```diff
@@ -0,0 +1,63 @@
+            if (is_nullable[i])
+            {
+                const ColumnNullable & nullable_col = assert_cast<const ColumnNullable &>(*columns[i]);
+                nested_columns[i] = &nullable_col.getNestedColumn();
```
How exactly it will work?
It looks like that we just took inner values of Nullable column regardless to null map.

alexey-milovidov replied:
What is out intent?
- to assume that NULL is an event but event does not match to any pattern;
- skip NULLs as no event exists.

alexey-milovidov replied:
If second, the implementation can be more simple...

alexey-milovidov replied:
Do I understand correctly that we know that the column is UInt8 and NULL is interpreted as 0?

achimbab replied:
My intent is to assume that NULL is an event.

To help explain my intent, I prepared data and a select query.

```SQL
create table funnel_test (dt DateTime, user int, event Nullable(String), action Nullable(String)) engine = Log()
insert into funnel_test values (1, 1, 'e1', NULL) (2, 1, NULL, 'a1')
```
```
                 dt   user   event   action

1970-01-01 09:00:01      1   e1      NULL  
1970-01-01 09:00:02      1   NULL    a1    
```

To find the events and actions flow of `event='e1' --> action='a1'`, let's execute below query.

```SQL
select user, windowFunnel(86400)(dt, event='e1', action='a1') from funnel_test group by user
```

But the result of query is NULL without this PullRequest.

And I found that `AggregateFunctionNull` skips any row containing a null-column-value.

So I implemented `AggregateFunctionWindowFunnelNoNullifyVariadic` and use values without null-map.

If there is any problems, let me know please.

Thank you for your comment.

alexey-milovidov commented on dbms/src/AggregateFunctions/AggregateFunctionNull.cpp:
```diff
@@ -51,10 +52,15 @@ class AggregateFunctionCombinatorNull final : public IAggregateFunctionCombinato
+            /// - that means - count number of calls, when all arguments are not NULL.
+            if (nested_function->getName() == "count")
+                return std::make_shared<AggregateFunctionCountNotNullUnary>(arguments[0], params);
+            else if (nested_function->getName() == "windowFunnel")
```
It looks like unnesessary dependency... it makes code looking weird.

achimbab replied:
@alexey-milovidov 
At the above scenario, I don't know find the events and actions flow of `event='e1' --> action='a1'` without AggregateFunctionWindowFunnelNoNullifyVariadic.
If you give me a little hint, I will try to fix it.
Thank you.

achimbab commented:
It is merged at https://github.com/ClickHouse/ClickHouse/pull/9938.
